pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -175,6 +175,11 @@ func (rc *reconciler) reconcile() {
 				continue
 			}
 
+			if rc.attacherDetacher.IsDetachBackoffError(attachedVolume.AttachedVolume) {
+				klog.V(5).Infof("Volume %v on node %v is not needed detaching until next backoff retry", attachedVolume.VolumeName, attachedVolume.NodeName)
+				continue
+			}
+
```

Comment:
From my POV, detaching volume cannot be "reverted" once it is triggered, even though it fails. Because VolumeAttachment Objected has been updated with DeletionTimestamp not nil, which will lead to [csi-attacher](https://github.com/kubernetes-csi/external-attacher) controller retry to detach volume from node until success. I think your solution can make sure node.status.volumesAttached is right, but how can you cancel csi-attacher controller retry to detach volume?

Reply:
Thanks for yongr attention. In my pr, we expect detaching volume normally，so the retry by csi-attacher is necessary，we do not need cancelling the retry.

Reply:
Thanks for yongr attention. In my pr, we expect detaching volume normally，so the retry by csi-attacher is necessary，we do not need cancelling the retry.

the reason of the issue, in a brief：

Reply:
Thanks for yongr attention. In my pr, our purpose is detaching volume normally, then re-attching successfully, not  cancelling the retry by controller. csi is also a volume plugin, so I think the retry by csi-attacher is necessary.

the reason of the issue, in a brief：

- first, failed to detach volume1 from node1, attach/detach controller will retry detaching volume1 after the backoff time
- second, the pod(or same volume) is scheduled to the same node(node1) , attach/detach controller will check volume1 if need detach and attach; because of watching a new pod, kubelet will verify the same volume(volume1) if attached.
- after comparing the data between asw and dsw ，attach/detach controller think volume1  not need detach and re-attach. But, volume1 is not detached successfully，then re-attach. So，the node.status.VolumesAttached do not have volume1
- when kubelet verify volume1 if attached, can not get volume1  from node.status.VolumesAttached. In result, pod will keep ContainerCreating

You can refrence this link for details： https://mp.weixin.qq.com/s/CElaEP5YCR3QF4ppOOicAw

Reply:
I can understand your question, but I think you didn't get my point. 
Once volume is triggered, you must wait it success and re-attach again, but I think your solution cannot make sure trigger re-attach operation at next attach-detach-controller reconcile. You just recover node.status.VolumesAttached but not stop csi-attacher retry to detach volume background, so there maybe appear  csi-attacher is detaching volume at background but new pod has startup by kubelet.

Reply:
What's more, I think attach-detach-controller can self-recovery after volume is successfully detached in this scenario. The recovery steps as following:
1. delete old pod, volume detachment triggered by attach-detach-controller
2. csi-attacher failed to detach volume, and attach-detach-controller reconcile retry it but meet backoff error, which lead 'node.status.VolumesAttached' is not fit with asw
3. new pod scheduled on same node, but cannot startup by kubelet because of 'node.status.VolumesAttached'
4. csi-attacher retry to detach volume util success, but csi-attacher cannot change attach-detach-controller's asw after detachement success 
5.  attach-detach-controller periodically check loop [func (rc *reconciler) syncStates()] will change its asw after volume is detached
6. attach-detach-controller trigger to attach volume via comparing asw and dsw, and add volume into 'node.status.VolumesAttached' again
7. new pod startup after several times retry

Reply:
看了下这个问题应该是csi-external-attacher 0.3 版本有bug导致的问题？而不是ADC的bug。